AMDGPU/GFX12: Insert waitcnts before stores with scope_sys

petar-avramovic opened the pull request:
Insert waitcnts for loads and atomics before stores with system scope.
Scope is field in instruction encoding and corresponds to desired
coherence level in cache hierarchy.
Intrinsic stores can set scope in cache policy operand.
If volatile keyword is used on generic stores memory legalizer will set
scope to system. Generic stores, by default, get lowest scope level.
Waitcnts are not required if it is guaranteed that memory is cached.
For example vulkan shaders can guarantee this.
TODO: implement flag for frontends to give us a hint not to insert waits.
Expecting vulkan flag to be implemented as vulkan:private MMRA.

jayfoad commented on llvm/lib/Target/AMDGPU/SIMemoryLegalizer.cpp:
```diff
@@ -2467,6 +2501,10 @@ bool SIMemoryLegalizer::expandStore(const SIMemOpInfo &MOI,
+
+  // GFX12 specific, scope(desired coherence domain in cache hierarchy) is
+  // instruction field, do not confuse it with atomic scope.
+  Changed |= CC->expandSystemScopeStore(MI);
```
Is this also needed for atomic stores? They returned early on line 2495 so they won't hit this code.

Pierre-vh replied:
I don't see why it wouldn't be needed on a store release? store release already waits, but it doesn't (explicitly) wait for all the counters we need to wait on for `expandSystemScopeStore`

Pierre-vh commented on llvm/test/CodeGen/AMDGPU/wait-for-stores-with-scope_sys.ll:
```diff
@@ -0,0 +1,16 @@
+; RUN: llc -march=amdgcn -mcpu=gfx1200 -verify-machineinstrs < %s | FileCheck -check-prefix=GFX12 %s
+; RUN: llc -global-isel -march=amdgcn -mcpu=gfx1200 -verify-machineinstrs < %s | FileCheck -check-prefix=GFX12 %s
+
+define amdgpu_ps void @intrinsic_store_system_scope(i32 %val, <4 x i32> inreg %rsrc, i32 %vindex, i32 %voffset, i32 inreg %soffset) {
```
https://llvm.org/docs/AMDGPUUsage.html#memory-scopes says the default is system scope, do simple `store i32 42, ptr %ptr` have SCOPE_SYS? If so they need to be tested as well

petar-avramovic replied:
Maybe that is default scope for atomics?
This is about gfx12 scope instruction field, default is 0 (SCOPE_CU) and generic `store i32 42, ptr %ptr` gets selected with SCOPE_CU (not printed in asm string). There is currently no way to set different "gfx12 scope" on generic stores.

arsenm commented on llvm/lib/Target/AMDGPU/SIMemoryLegalizer.cpp:
```diff
@@ -2381,6 +2387,34 @@ bool SIGfx12CacheControl::enableVolatileAndOrNonTemporal(
+    // For example shader-visible memory is cached.
+    // TODO: implement flag for frontend to give us a hint not to insert waits.
+    MachineBasicBlock &MBB = *MI->getParent();
+    DebugLoc DL = MI->getDebugLoc();
```
const ref

Pierre-vh commented on llvm/lib/Target/AMDGPU/SIMemoryLegalizer.cpp:
```diff
@@ -2381,6 +2387,34 @@ bool SIGfx12CacheControl::enableVolatileAndOrNonTemporal(
+    MachineBasicBlock::iterator &MI) const {
+
+  MachineOperand *CPol = TII->getNamedOperand(*MI, OpName::cpol);
+  if (CPol && ((CPol->getImm() & CPol::SCOPE) == CPol::SCOPE_SYS)) {
```
invert condition & early return

petar-avramovic committed:
AMDGPU/GFX12: Insert waitcnts before stores with scope_sys

Insert waitcnts for loads and atomics before stores with system scope.
Scope is field in instruction encoding and corresponds to desired
coherence level in cache hierarchy.
Intrinsic stores can set scope in cache policy operand.
If volatile keyword is used on generic stores memory legalizer will set
scope to system. Generic stores, by default, get lowest scope level.
Waitcnts are not required if it is guaranteed that memory is cached.
For example vulkan shaders can guarantee this.
TODO: implement flag for frontends to give us a hint not to insert waits.
Expecting vulkan flag to be implemented as vulkan:private MMRA.

petar-avramovic commented:
Volatile stores set scope_sys, insert waits there also

jayfoad commented on llvm/lib/Target/AMDGPU/SIMemoryLegalizer.cpp:
```diff
@@ -2364,6 +2396,9 @@ bool SIGfx12CacheControl::enableVolatileAndOrNonTemporal(
     Changed |= setScope(MI, AMDGPU::CPol::SCOPE_SYS);
 
+    if (Op == SIMemOp::STORE)
+      Changed |= insertWaitsBeforeSystemScopeStore(MI);
```
It is a bit messy that we need this extra call to insertWaitsBeforeSystemScopeStore here, because the call to insertWait below modifies MI so it no longer refers to the store. But I guess it is OK.

jayfoad commented on llvm/lib/Target/AMDGPU/SIMemoryLegalizer.cpp:
```diff
@@ -589,6 +593,16 @@ class SIGfx12CacheControl : public SIGfx11CacheControl {
+  //   TODO: SIInsertWaitcnts will not always be able to remove STORECNT waits
+  //   since it does not distinguish atomics-with-return from regular stores.
+  // There is no need to wait if memory is cached (mtype != UC).
+  // For example shader-visible memory is cached.
```
I don't understand the statement that "shader-visible memory is cached". Surely we are compiling a shader, so any memory the shader refers to is "shader-visible", so why do we need to worry about uncached memory?

jayfoad commented on llvm/lib/Target/AMDGPU/SIMemoryLegalizer.cpp:
```diff
@@ -2381,6 +2416,15 @@ bool SIGfx12CacheControl::enableVolatileAndOrNonTemporal(
+    MachineBasicBlock::iterator &MI) const {
+  MachineOperand *CPol = TII->getNamedOperand(*MI, OpName::cpol);
+  if (!CPol || ((CPol->getImm() & CPol::SCOPE) != CPol::SCOPE_SYS))
+    return false;
```
Returning early no longer makes this code any shorter or clearer.
